### DISCUSSION TITLE
supabase key is required (when pass service role key as 2nd param in createClient)

### DISCUSSION THREAD

**Discussion by ousszizou** [supabase/supabase#3919] [Questions]
I added service role key to .env.local file ( Next.js Project ) & pass it to createClient() as below.

If I use the syntax : "process.env.SUPABASE_SERVICE_ROLE_KEY" it throw the error: **supabase key is required.** I checked the value of process.env.SUPABASE_SERVICE_ROLE_KEY in the console & it exists. I don't know why **createClient** ignore my supabaseKey.
```
export const supabaseAdmin = createClient(
  process.env.NEXT_PUBLIC_SUPABASE_URL as string,
  process.env.SUPABASE_SERVICE_ROLE_KEY as string
);

```

But If I use it directly as string value, it works fine & not throw any error.

```
export const supabaseAdmin = createClient(
  process.env.NEXT_PUBLIC_SUPABASE_URL as string,
"abJhbGciOiJIUzI1nR5cCI6IkpXVCJ9.eyJyb2xlIjoic2VydmljZV9yb2xlIiwiaWF0IjoxNjzB9.piUWg0BEMhSeOo4Eb2h0XAdnMcYR5aEY5V8Pkc"
);
```

my .env.local looks like that:

```
NEXT_PUBLIC_SUPABASE_URL=https://my-project.supabase.co
NEXT_PUBLIC_SUPABASE_ANON_KEY= key value
SUPABASE_SERVICE_ROLE_KEY= key value
```

**Comment by silentworks** [comment]
Can you console.log the `process.env.SUPABASE_SERVICE_ROLE_KEY` in your code to make sure it is present. I'm starting to think it isn't present in the context of Next.js when it's running.

**Reply by ousszizou** [reply]
I checked the value of process.env.SUPABASE_SERVICE_ROLE_KEY by console.log & it exists.

**Reply by silentworks** [reply]
Then there shouldn't be any reason for the error. You are console logging it inside the same file you are trying to use it, also is their being logged to the terminal or the browser? because service role should only be used server-side and not client-side.

**Reply by ousszizou** [reply]
I tried the same here : https://github.com/vercel/nextjs-subscription-payments/blob/main/utils/supabase-admin.ts

why above works fine. but in my case NO. 

> also is their being logged to the terminal or the browser?

IN BROWSER

> because service role should only be used server-side and not client-side.

Yes, I'll use it in server-side but I didn't use it yet cuz I can't initialize supabaseAdmin by createClient()

**Reply by kangmingtay** [reply]
You need to prefix the env var with NEXT_PUBLIC

**Comment by hynekfisera** [comment]
I'm having exactly the same problem. Did you manage to find a solution?

**Reply by ousszizou** [reply]
Unfortunately NO.

**Reply by kangmingtay** [reply]
If i'm not mistaken, since this is a nextjs app, your env var has to be prefixed with NEXT_PUBLIC so it shd look like NEXT_PUBLIC_SERVICE_KEY. Hope this helps!

**Reply by 0xAskar** [reply]
i have this already, but it didn't fix my problem

**Reply by kangmingtay** [reply]
@0xAskar it will be great if you can describe the steps to reproduce the error you're facing so it's easier for the community to help solve your issue. As a sanity check, i would try to see if i can read those environment variables via console logging and check that there aren't any extra quotes / typo errors in your env var.

**Reply by 0xAskar** [reply]
It seems like my error was that the api key for the provider was incorrect. so in my case, i was using twitter as the provider and the apikey I provided on the Supabase site for my project were wrong. So in case anyone has that error, I would advise double checking your provider's credentials as well

**Reply by nicitaacom** [reply]
NO - don't do it
NEXT_PUBLIC_ means its may be exposed to client
You don't want to expose your SERVICE_ROLE_KEY to client - it must be always available ONLY on server

If you mean NEXT_PUBLIC_SUPABASE_URL - its fine I agree you had to add NEXT_PUBLIC to it

**Reply by indigoscipio** [reply]
hi, so should i put inside next public or not?

**Reply by kangmingtay** [reply]
@indigoscipio no, you should never expose the service role on the client side - so don't put it in `NEXT_PUBLIC_`

**Comment by nicitaacom** [comment]
Today I got the same issue

Somebody found solution for that?
Yesterday it worked find - today something happened

Also I want mention that it happens only on dynamic route (in my case [ticketId]) - in other API routes everything works fine (I mean when user from / route send API call)

Update: It happens when I use react-hook-form - without this works fine

**Comment by rookledookle** [comment]
Facing the same issue

**Reply by kangmingtay** [reply]
@rookledookle have you sanity checked that the env var can be read from wherever you are calling createClient from? we haven't made any changes to the instantiation of the supabase client recently

**Comment by reghope** [comment]
Also having the same issue

**Reply by kangmingtay** [reply]
@bookerlyio have you sanity checked that the env var can be read from wherever you are calling `createClient` from? we haven't made any changes to the instantiation of the supabase client recently

**Reply by reghope** [reply]
> @bookerlyio have you sanity checked that the env var can be read from wherever you are calling `createClient` from? we haven't made any changes to the instantiation of the supabase client recently

Thanks for the reply - managed to track it down just before it's called, and it's blank. Can't figure out why. It's set properly in AWS Amplify and other env variables work. Here's how I was checking:

```
const supabaseURL = process.env.NEXT_PUBLIC_SUPABASE_URL || '';
const supabaseKey = process.env.SUPABASE_SERVICE_ROLE_KEY || '';

console.log('supabaseURL'  + supabaseURL);
console.log('supabaseKey ' + supabaseKey);

const supabaseAdmin = createClient<Database>(supabaseURL, supabaseKey);
```

**Reply by reghope** [reply]
Fixed:

For anyone on AWS Amplify, you need to 'declare' your environment variables in a amplify.yml in your root directory, and also put them in your environment variables/secrets in AWS amplify too. (https://docs.aws.amazon.com/amplify/latest/userguide/ssr-environment-variables.html)
